bullet_train/lib/bullet_train/core_ext/callbacks_instrumentation.rb

```diff
@@ -0,0 +1,58 @@
+require_relative "callbacks_instrumentation/backport"
+
+::ActiveSupport::Callbacks::ClassMethods.prepend Module.new {
+  def set_callback(name, *filter_list, &block)
+    *, options = normalize_callback_params(filter_list.dup, block)
+    Instrumenter.options = options.merge(caller_locations:)
+    super
+  ensure
+    Instrumenter.options = nil
+  end
+}
+
+::ActiveSupport::Callbacks::Callback.prepend Module.new {
+  def compiled
+    @compiled = super.then { Instrumenter.build self, _1 } unless defined?(@compiled)
+    @compiled
+  end
+
+  class Instrumenter < Data.define(:callback, :callable, :options)
+    singleton_class.attr_accessor :options
+
+    def self.build(callback, callable)
+      options ? new(callback, callable, options) : callable
+    end
+
+    def initialize(callback:, callable:, options:)
+      caller_locations = options.delete(:caller_locations)
+      @location  = Rails.backtrace_cleaner.clean(caller_locations).first if caller_locations
+      @full_name = callback.name == :commit ? caller_locations.first.base_label : "#{callback.kind}_#{callback.name}"
+      super
+    end
+    attr_reader :options, :location, :full_name
+    delegate :name, :kind, :filter, to: :callback
+
+    def apply(sequence)
+      case kind
+      in :before then sequence.before(self)
+      in :after  then sequence.after(self)
+      in :around then sequence.around(self)
+      end
+    end
+
+    def call(env)
+      ActiveSupport::Notifications.instrument "callback.active_support", callback: self do
+        callable.call(env)
+      end
+    end
+  end
+}
+
+class ActiveSupport::Callbacks::LogSubscriber < ActiveSupport::LogSubscriber
+  attach_to :active_support
+
+  def callback(event)
+    event.payload => {callback:}
+    info { "Callback #{callback.full_name}: #{callback.filter} #{callback.options} from #{callback.location}".squish }
```

Comment:
Here's a sample of what it looks like in practice:

```
irb(main):001> User.first.touch
  User Load (1.5ms)  SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT $1  [["LIMIT", 1]]
Callback after_find: after_find_invoked from app/models/user.rb:27:in `invoke'
Callback after_initialize: after_initialize_invoked from app/models/user.rb:27:in `invoke'
  TRANSACTION (0.1ms)  BEGIN
  User Update (1.1ms)  UPDATE "users" SET "updated_at" = $1 WHERE "users"."id" = $2  [["updated_at", "2024-01-29 20:41:18.519670"], ["id", 1]]
Callback after_touch: after_touch_invoked from app/models/user.rb:27:in `invoke'
  TRANSACTION (0.5ms)  COMMIT
Callback after_destroy_commit: after_destroy_commit_invoked from app/models/user.rb:27:in `invoke'
Callback after_update_commit: after_update_commit_invoked from app/models/user.rb:27:in `invoke'
Callback after_save_commit: after_save_commit_invoked from app/models/user.rb:27:in `invoke'
Callback after_create_commit: after_create_commit_invoked from app/models/user.rb:27:in `invoke'
Callback after_commit: #<Proc:0x000000010c97bdc0 /Users/kasperhansen/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/activestorage-7.1.2/lib/active_storage/attached/model.rb:127> from /Users/kasperhansen/code/bullet-train-co/bullet_train-core/bullet_train/app/models/concerns/users/base.rb:33:in `block in <module:Base>'
[ActionCable] Broadcasting to User: {:changed=>["updated_at"]}
[ActionCable] Broadcasting to gid://untitled/User/1: {:changed=>["updated_at"]}
=> true
```

Note: I'm not sure why `after_destroy_commit` is invoked on `touch` but it looked like a Rails issue, IMO.

`invoke` was just a helper method to schedule all the callbacks:

```ruby
class User
  def self.invoke(callback, with: :"#{callback}_invoked")
    define_method(with) { yield if callback.to_s.include?("around") }
    send(callback, with)
  end

  invoke :after_initialize
  invoke :after_find
  invoke :after_touch

  before_validation { puts "before_validation" }
  before_validation { throw :abort }
  before_validation { puts "after before_validation" }

  invoke :before_validation
  invoke :after_validation

  invoke :around_create
  invoke :before_create
  invoke :after_create

  invoke :around_save
  invoke :before_save
  invoke :after_save

  invoke :around_update
  invoke :before_update
  invoke :after_update

  invoke :around_destroy
  invoke :before_destroy
  invoke :after_destroy

  invoke :after_create_commit
  invoke :after_save_commit
  invoke :after_update_commit
  invoke :after_destroy_commit
end
```